pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,220 @@
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different with the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
+	result := make(schedulerapi.HostPriorityList, len(nodes))
+	// return if incoming pod doesn't have soft topology spread constraints.
+	constraints := getSoftTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return result, nil
+	}
+
+	t := newTopologySpreadConstraintsMap(len(nodes))
+	t.initialize(pod, nodes)
+
+	allNodeNames := make([]string, 0, len(nodeNameToInfo))
+	for name := range nodeNameToInfo {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	errCh := schedutil.NewErrorChannel()
+	ctx, cancel := context.WithCancel(context.Background())
+	processAllNode := func(i int) {
+		nodeInfo := nodeNameToInfo[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			return
+		}
+		// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
+		// (2) All topologyKeys need to be present in `node`
+		if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+			!predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			return
+		}
+
+		for _, constraint := range constraints {
+			pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+			// If current topology pair is not associated with any "filtered" node,
+			// continue to avoid unnecessary calculation.
+			if t.topologyPairToPodCounts[pair] == nil {
+				continue
+			}
+
+			// <matchSum> indicates how many pods (on current node) match the <constraint>.
+			matchSum := int64(0)
+			for _, existingPod := range nodeInfo.Pods() {
+				match, err := predicates.PodMatchesSpreadConstraint(existingPod.Labels, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if match {
+					matchSum++
+				}
+			}
+			atomic.AddInt64(t.topologyPairToPodCounts[pair], matchSum)
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processAllNode)
+	if err := errCh.ReceiveError(); err != nil {
+		return nil, err
+	}
+
+	var minCount int64 = math.MaxInt64
+	// <total> sums up the number of matching pods on each qualified topology pair
+	var total int64
+	for _, node := range nodes {
+		if t.nodeNameToPodCounts[node.Name] == -1 {
+			continue
+		}
+
+		// For each present <pair>, current node gets a credit of <matchSum>.
+		// And we add <matchSum> to <t.total> to reverse the final score later.
+		for _, constraint := range constraints {
+			if tpVal, ok := node.Labels[constraint.TopologyKey]; ok {
+				pair := topologyPair{key: constraint.TopologyKey, value: tpVal}
+				matchSum := atomic.LoadInt64(t.topologyPairToPodCounts[pair])
```

Comment:
This is not running in parallel, so no need to use atomic. Same below.

Reply:
This case is a _little_ tricky b/c we need to **concurrently** compare with `minCount`, and it's a racing behavior. We _may_ need to use `atomic.CompareAndSwapInt64`. Will give a try on this.

Reply:
I did a test ([gist](https://gist.github.com/Huang-Wei/7fa9eafb7074e3b4556211aa4723afe4)) and seems parallel version doesn't help much.